test/util.ts

```diff
@@ -0,0 +1,119 @@
+          // actual.e === expected.e we can clean out the expecteds that no one can claim
+          // anymore.
+          if(actual.equal(expected, true /*ignore the node*/, true /*ignore the e*/)) {
+            found = true;
+            potentials.push({e: expected.e, relatedChanges: [expectedIx]});
+          }
+          expectedIx++;
+        }
+        // if there was only one match, no one can ever have this expected - we've claimed it.
+        // As such, we need to remove it from the list;
+        if(potentials.length === 1) {
+          all.splice(potentials[0].relatedChanges[0], 1);
+        }
+        eMap[actual.e] = potentials;
+      } else if(matches.length === 1) {
+        // in the case where we've mapped our actual.e to our expected.e, we just check this
+        // current fact for a match where the expected.e is what we're looking for
+        for(let expected of all) {
+          if(expected.e === matches[0].e && actual.equal(expected, true, true)) {
+            found = true;
+            all.splice(expectedIx, 1);
+          }
+          expectedIx++;
+        }
+      } else {
+        // since we have multiple potential matches, we need to see if this actual might reduce
+        // the set down for us. For each expected that's left, we'll check if expected.e matches
+        // one of our potentials, and if this expected would equal our actual if we ignored the
+        // e. If so, we keep this potential in the running. Any potentials the don't end up with
+        // a match get removed on account of us recreating the potential array from scratch here.
+        let potentials = [];
+        for(let expected of all) {
+          for(let match of matches) {
+            if(match.e === expected.e && actual.equal(expected, true, true)) {
+              found = true;
+              potentials.push(match);
+              match.relatedChanges.push(expectedIx);
+            }
+          }
+          expectedIx++;
+        }
+        // If we only have one potential, we need to clean up after ourselves again. This time
+        // however, we could have had many relatedChanges that we need to clean up, so we'll loop
+        // through them and remove them from the list. They're our's now.
+        if(potentials.length === 1) {
+          let related = potentials[0].relatedChanges;
+          related.sort((a:number, b:number) => b - a);
+          for(let relatedIx of related) {
+            all.splice(relatedIx, 1);
+          }
+          potentials[0].relatedChanges = []
+        }
+        eMap[actual.e] = potentials;
+      }
+
+      if(!found) assert.fail("No match found for: " + actual.toString());
+      else assert.pass("Found match for: " + actual.toString());
+    }
+  }
+}
+
+export function time(start?:any): number | number[] | string {
+  if ( !start ) return process.hrtime();
+  let end = process.hrtime(start);
+  return ((end[0]*1000) + (end[1]/1000000)).toFixed(3);
```

Comment:
isn't this `(end[1]/1000000)` off by an order of magnitude? or is lack of sleep getting to me?

> The process.hrtime() method returns the current high-resolution real time in a [seconds, nanoseconds] tuple Array.

1ms = 10e6ns